File: generated_code/GeminiPro/Python/071_SplayTree.py
import sys

class SplayTree:
    """
    An implementation of a Splay Tree that acts as a dictionary-like set for integers.
    A splay tree is a self-balancing binary search tree with the additional property
    that recently accessed elements are quick to access again.
    """

    class _Node:
        """A node in the splay tree."""
        def __init__(self, key, parent=None):
            self.key = key
            self.parent = parent
            self.left = None
            self.right = None

    def __init__(self):
        """Initializes an empty SplayTree."""
        self.root = None

    def _left_rotate(self, x):
        """Performs a left rotation on node x."""
        y = x.right
        if y:
            x.right = y.left
            if y.left:
                y.left.parent = x
            y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y
        if y:
            y.left = x
        x.parent = y

    def _right_rotate(self, x):
        """Performs a right rotation on node x."""
        y = x.left
        if y:
            x.left = y.right
            if y.right:
                y.right.parent = x
            y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.right:
            x.parent.right = y
        else:
            x.parent.left = y
        if y:
            y.right = x
        x.parent = y

    def _splay(self, node):
        """
        Performs the splay operation on a node, moving it to the root.
        """
        if not node:
            return
        while node.parent:
            parent = node.parent
            grandparent = parent.parent
            if not grandparent:  # Zig case
                if node == parent.left:
                    self._right_rotate(parent)
                else:
                    self._left_rotate(parent)
            elif node == parent.left and parent == grandparent.left:  # Zig-Zig (left-left)
                self._right_rotate(grandparent)
                self._right_rotate(parent)
            elif node == parent.right and parent == grandparent.right:  # Zig-Zig (right-right)
                self._left_rotate(grandparent)
                self._left_rotate(parent)
            elif node == parent.right and parent == grandparent.left:  # Zig-Zag (left-right)
                self._left_rotate(parent)
                self._right_rotate(grandparent)
            else:  # Zig-Zag (right-left)
                self._right_rotate(parent)
                self._left_rotate(grandparent)
        self.root = node

    def search(self, key):
        """
        Searches for a key in the tree.
        
        Performs the splaying operation on the accessed node if found,
        or on the last accessed node on the search path if not found.
        
        Args:
            key: The integer key to search for.
            
        Returns:
            True if the key is found, False otherwise.
        """
        if not self.root:
            return False

        current = self.root
        last_visited = self.root
        while current:
            last_visited = current
            if key < current.key:
                current = current.left
            elif key > current.key:
                current = current.right
            else:
                # Key found, splay it and return True
                self._splay(current)
                return True

        # Key not found. Splay the last node on the search path.
        self._splay(last_visited)
        return False

    def insert(self, key):
        """
        Inserts a key into the tree.
        
        If the key already exists, the node with that key is splayed to the root.
        If the key does not exist, it is inserted and the new node is splayed to the root.
        
        Args:
            key: The integer key to insert.
        """
        if not self.root:
            self.root = self._Node(key)
            return

        # Splay the closest node to the root. This handles the existing key case.
        if self.search(key):
            return  # Key already exists, search has splayed it.

        # After search, the root is the closest node to where the key should be.
        new_node = self._Node(key)
        if key < self.root.key:
            new_node.right = self.root
            new_node.left = self.root.left
            if self.root.left:
                self.root.left.parent = new_node
            self.root.left = None
        else:  # key > self.root.key
            new_node.left = self.root
            new_node.right = self.root.right
            if self.root.right:
                self.root.right.parent = new_node
            self.root.right = None
        
        self.root.parent = new_node
        self.root = new_node

    def _find_max(self, node):
        """Finds the node with the maximum key in a subtree."""
        if not node:
            return None
        current = node
        while current.right:
            current = current.right
        return current

    def delete(self, key):
        """
        Deletes a key from the tree.
        
        If the key is found, it is removed, and its parent or a replacement
        node is splayed. If not found, the last node on the search path is splayed.
        
        Args:
            key: The integer key to delete.
        """
        # Splay the node to the root. If not found, splay the last visited node.
        if not self.search(key):
            return  # Key not found, search already splayed the closest node.

        # At this point, the node to delete is the root.
        left_subtree = self.root.left
        right_subtree = self.root.right

        if not left_subtree:
            # Promote the right subtree
            self.root = right_subtree
            if right_subtree:
                right_subtree.parent = None
        else:
            # Disconnect the left subtree to work with it independently
            left_subtree.parent = None
            
            # Find the maximum element in the left subtree
            max_in_left = self._find_max(left_subtree)
            
            # Splay this max element to make it the root of the left subtree
            # Temporarily set the root to perform the splay within the subtree
            original_root = self.root
            self.root = left_subtree
            self._splay(max_in_left)
            
            # The new root (max_in_left) now has no right child.
            # Attach the original right subtree to it.
            self.root.right = right_subtree
            if right_subtree:
                right_subtree.parent = self.root
            
            # Clean up references from the deleted node (optional, for GC)
            original_root.left = None
            original_root.right = None

File: generated_code/GeminiPro/Java/071_SplayTree.java
/**
 * A self-contained Java class implementing a Splay Tree for a set of integers.
 * A Splay Tree is a self-balancing binary search tree with the additional property
 * that recently accessed elements are quick to access again. It performs basic
 * operations such as insertion, deletion, and search. After any of these operations,
 * the accessed node is moved to the root of the tree through a process called "splaying".
 */
public class SplayTree {

    /**
     * Represents a node in the Splay Tree.
     * This is a private static inner class to encapsulate the node structure.
     */
    private static class Node {
        int key;
        Node left, right;

        /**
         * Constructs a new node with the given key.
         *
         * @param key The integer key for the node.
         */
        Node(int key) {
            this.key = key;
            this.left = null;
            this.right = null;
        }
    }

    private Node root;

    /**
     * Constructs an empty Splay Tree.
     */
    public SplayTree() {
        this.root = null;
    }

    /**
     * The core splaying operation. Moves the node with the given key, or the last
     * accessed node if the key is not found, to the root of the tree.
     * This implementation uses the top-down splaying technique.
     *
     * @param key The key to splay.
     */
    private void splay(int key) {
        if (root == null) {
            return;
        }

        // Dummy node to simplify linking. header.right will be the root of the
        // "left" tree (elements smaller than the key) and header.left will be
        // the root of the "right" tree (elements larger than the key).
        Node header = new Node(0); // Dummy node, key value is irrelevant.
        Node leftTreeMax = header;
        Node rightTreeMin = header;

        Node current = root;

        while (true) {
            if (key < current.key) {
                if (current.left == null) {
                    break;
                }
                // Zig-Zig case (Right Rotation)
                if (key < current.left.key) {
                    Node temp = current.left;
                    current.left = temp.right;
                    temp.right = current;
                    current = temp;
                    if (current.left == null) {
                        break;
                    }
                }
                // Link to the right tree
                rightTreeMin.left = current;
                rightTreeMin = current;
                current = current.left;
            } else if (key > current.key) {
                if (current.right == null) {
                    break;
                }
                // Zig-Zig case (Left Rotation)
                if (key > current.right.key) {
                    Node temp = current.right;
                    current.right = temp.left;
                    temp.left = current;
                    current = temp;
                    if (current.right == null) {
                        break;
                    }
                }
                // Link to the left tree
                leftTreeMax.right = current;
                leftTreeMax = current;
                current = current.right;
            } else {
                // Key found
                break;
            }
        }

        // Reassemble the tree
        leftTreeMax.right = current.left;
        rightTreeMin.left = current.right;
        current.left = header.right;
        current.right = header.left;

        root = current;
    }

    /**
     * Searches for a key in the tree. If the key is found, it is splayed
     * to the root. If not found, the last accessed node is splayed to the root.
     *
     * @param key The key to search for.
     * @return The key if found, otherwise null.
     */
    public Integer search(int key) {
        if (root == null) {
            return null;
        }
        splay(key);
        if (root.key == key) {
            return key;
        }
        return null;
    }

    /**
     * Inserts a key into the tree. Duplicates are not allowed.
     * The newly inserted node becomes the new root of the tree.
     *
     * @param key The key to insert.
     */
    public void insert(int key) {
        // If the tree is empty, create a new root.
        if (root == null) {
            root = new Node(key);
            return;
        }

        // Splay the tree for the key.
        splay(key);

        // If the key is already present, do nothing.
        if (root.key == key) {
            return;
        }

        Node newNode = new Node(key);

        // The new node becomes the root. The old root becomes a child.
        if (key < root.key) {
            newNode.left = root.left;
            newNode.right = root;
            root.left = null;
        } else { // key > root.key
            newNode.right = root.right;
            newNode.left = root;
            root.right = null;
        }
        root = newNode;
    }

    /**
     * Deletes a key from the tree.
     * After deletion, the parent of the deleted node (or a nearby node) is splayed to the root.
     *
     * @param key The key to delete.
     */
    public void delete(int key) {
        if (root == null) {
            return;
        }

        // Splay the tree for the key. If the key exists, it becomes the root.
        splay(key);

        // If the key is not at the root, it's not in the tree.
        if (root.key != key) {
            return;
        }

        // Now, the node to be deleted is the root.
        Node leftSubtree = root.left;
        Node rightSubtree = root.right;

        if (leftSubtree == null) {
            // If there's no left subtree, the right subtree becomes the new tree.
            root = rightSubtree;
        } else {
            // Make the left subtree the main tree.
            root = leftSubtree;
            // Splay the left subtree for the key 'key'. This will bring the largest
            // element in the left subtree (the predecessor of the deleted key)
            // to the root of this new tree.
            splay(key);

            // The new root (the max of the old left subtree) has no right child.
            // Attach the original right subtree to it.
            root.right = rightSubtree;
        }
    }
}